proposals/compound-extensions.md

```diff
@@ -0,0 +1,380 @@
+# Compound Extensions
+
+* Type: Design proposal
+* Author: Chuck Jazdzewski
+* Status: proposed
+* Prototype: none
+* Discussion: tbd
+
+## Summary
+
+Extensions can be used to extend a type with an extension.
+
+## Introduction
+
+### Description
+
+Compound extensions are extensions that introduce an extension onto another type.
+
+A compound extension is declared using dotted list of types. For example,
+
+```kotlin
+operator fun Body.String.unaryPlus(s: String) = escape(s)
```

Comment:
This can't work. What does `Map.Entry.unaryPlus` apply to? `Map` and `Entry` or `Map.Entry`?

Reply:
I think this is discussed later on in the syntax part of the proposal. As far as I can tell, this example uses the current syntax as an introduction to the topic. This syntax clearly does not work, neither with namespaces (as pointed out below) not nested classes. 
The question is which of the other possible syntax variants should be used primarily. I personally prefer parenthesis.

Reply:
It seems like square brackets give the best readability and parseability. We should remember that we need the syntax for function types. `[].()->` reads better than `().()->`. I think that alternative syntax solutions should be added to proposal.

Reply:
It works in either case as addressed below.

In the where there is an existing meaning then such as `Map.Entry` it refers to `Map.Entry`. If `Map` doesn't have a nested type named `Entry` it refers to the `Entry` in scope. If you have an `Entry` in scope you wish to extend, it collides with `Entry` but a alias can be introduced to allow you to disambiguate.

Reply:
@chuckjaz What about using the `+` character in place of the possibly ambiguous `.`?
The plus symbol on types at receiver declaration is currently not valid Kotlin, so it wouldn't break any code, and I think it'd be more clear than a dot.

Reply:
Example:
```kotlin
operator fun Body+String.unaryPlus(s: String) = escape(s)
```

Reply:
Can you create a pull request to my branch to incorporate this as an alternative syntax?

Reply:
> What about using the `+` character in place of the possibly ambiguous `.`?
> The plus symbol on types at receiver declaration is currently not valid Kotlin, so it wouldn't break any code, and I think it'd be more clear than a dot.

It is better, than dot, but readability is still worse than brackets. You need to remember, that the same notation for functional types. And there should be place for parameters (and parameters in theory could also have multi-receivers).